### PR TITLE
Refactor `MaterialStateProperty` lerp functions

### LINKED ISSUE #60555: Need a definition of OutlinedShape.lerp
OutlinedBorder was added a few months ago: https://github.com/flutter/flutter/pull/56341. Currently it does not include a lerp method.

### LINKED ISSUE #104506: Refactor `MaterialStateProperty` lerp functions
Currently `MaterialStateProperty` lerp functions are localized to each class and duplicated across the source code.

### `button_style.dart`

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/button_style.dart#L475

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/button_style.dart#L497

### `scrollbar_theme.dart`

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/scrollbar_theme.dart#L274

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/scrollbar_theme.dart#L287

### `checkbox_theme.dart`

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/checkbox_theme.dart#L188

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/checkbox_theme.dart#L208

### `data_table_theme.dart`

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/data_table_theme.dart#L189

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/data_table_theme.dart#L197

### `navigation_bar_theme.dart`

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/navigation_bar_theme.dart#L180

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/navigation_bar_theme.dart#L193

### `radio_theme.dart`

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/radio_theme.dart#L158

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/radio_theme.dart#L171

### `switch_theme.dart`

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/switch_theme.dart#L146

https://github.com/flutter/flutter/blob/5739bf4fabb8340ced93eecc0d15072498d3a7ac/packages/flutter/lib/src/material/switch_theme.dart#L159

### Solution

Remove duplication and add lerp functions to `material_state.dart`

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes https://github.com/flutter/flutter/issues/104506
fixes https://github.com/flutter/flutter/issues/60555

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by HansMuller — COMMENTED** [review]
This is a nice improvement. Enshrining the special cases for BorderSide and OutlinedBorder in new public API doesn't seem like a good idea; better to create lerp functions for those types.

**Comment by HansMuller** [comment]
By "better to create lerp functions for those types" I meant that it would be better to update BorderSide.lerp to handle nulls, and  ... it looks like you can already define MaterialStateProperty.lerp in terms of ShapeBorder.lerp already?

**Comment by TahaTesser** [comment]
>  By "better to create lerp functions for those types" I meant that it would be better to update BorderSide.lerp to handle nulls, and ... it looks like you can already define MaterialStateProperty.lerp in terms of ShapeBorder.lerp already?

Thank you, I understand it better now. 

`ShapeBorder.lerp` requires casting to OutlinedBorder  as a result I'm adding `OutlinedBorder.lerp` from https://github.com/flutter/flutter/issues/60555 and I will also update `BorderSide.lerp` to handle nulls.

File: packages/flutter/lib/src/painting/borders.dart [hunk 1]
```diff
@@ -566,6 +566,41 @@ abstract class OutlinedBorder extends ShapeBorder {
...
+  /// This defers to `b`'s [lerpTo] function if `b` is not null. If `b` is
+  /// null or if its [lerpTo] returns null, it uses `a`'s [lerpFrom]
+  /// function instead. If both return null, it returns `a` before `t=0.5`
+  /// and `b` after `t=0.5`.
+  ///
+  /// {@macro dart.ui.shadow.lerp}
+  static OutlinedBorder? lerp(OutlinedBorder? a, OutlinedBorder? b, double t) {
+    assert(t != null);
+    ShapeBorder? result;
+    if (b != null)
```
Comment (HansMuller) [inline packages/flutter/lib/src/painting/borders.dart:598]: We're moving towards requiring curly braces for all control flow statements - https://github.com/flutter/flutter/issues/104498 - best to apply that rule to this code now.

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by TahaTesser** [comment]
> Google testing — No tests has been executed due to internal error. Ask a Googler to investigate.

Trying again

**Comment by HansMuller** [comment]
I asked on hackers-infra, I don't understand why "Google Testing" keeps failing. The [details link](https://flutter-rob.corp.google.com/#/flutter/premerge_prs?pr=104507) just points to an empty FROB page.

**Comment by TahaTesser** [comment]
Thanks for raising this. (I cannot access the details without credentials)

**Comment by keyonghan** [comment]
It's not showing in [go/flutter-rob](http://goto.google.com/flutter-rob) as it had already been tested 8 times. @HansMuller 
 Feel free to land it if you are comfortable and we can verify it in postsubmit.

@CaseyHillers filed [b/234758732](https://b.corp.google.com/234758732) to track.

**Comment by HansMuller** [comment]
@keyonghan - OK, I'm going to land this. It should be safe.

**Pull request merged**